Add `cirq.SqrtIswapTargetGateset` for (parameterized & non-parameterized) compilation to sqrt iswaps.

tanujkhattar opened the pull request:
- Adds new `cirq.SqrtIswapTargetGateset` primitive.
- New primitive supports both analytical decomposition of non-parameterized gates and decomposition of parameterized primitives of fsim family (+ implicit cirq decompose)
- Part of https://github.com/quantumlib/Cirq/issues/4722
- Part of https://github.com/quantumlib/Cirq/issues/5028
- Follows the new Transformer API https://github.com/quantumlib/Cirq/issues/4483
- Supports no compile tags NoCompile Tag for optimizers https://github.com/quantumlib/Cirq/issues/4253
- Fixes https://github.com/quantumlib/Cirq/issues/4214


cc @MichaelBroughton @dstrain115

MichaelBroughton commented on cirq-core/cirq/transformers/target_gatesets/sqrt_iswap_gateset.py:
```diff
@@ -0,0 +1,281 @@
+        gate = op.gate
+        q0, q1 = op.qubits
+
+        if isinstance(gate, ops.CZPowGate):
+            return _cphase_symbols_to_sqrt_iswap(q0, q1, gate.exponent, self.use_sqrt_iswap_inv)
+        if isinstance(gate, ops.SwapPowGate):
+            return _swap_symbols_to_sqrt_iswap(q0, q1, gate.exponent, self.use_sqrt_iswap_inv)
+        if isinstance(gate, ops.ISwapPowGate):
+            return _iswap_symbols_to_sqrt_iswap(q0, q1, gate.exponent, self.use_sqrt_iswap_inv)
+        if isinstance(gate, ops.FSimGate):
+            return _fsim_symbols_to_sqrt_iswap(
+                q0, q1, gate.theta, gate.phi, self.use_sqrt_iswap_inv
+            )
```
Why are these extra cases needed here ? It looks like two_qubit_to_sqrt_iswap handles a lot of those (or could be made to much more cleanly there).

tanujkhattar replied:
These are cases for compiling parameterized gates. `two_qubit_matrix_to_sqrt_iswap_operations` does analytical decomposition of a two qubit matrix and hence will not work for parameterized gates.

MichaelBroughton commented on cirq-core/cirq/transformers/target_gatesets/sqrt_iswap_gateset.py:
```diff
@@ -0,0 +1,281 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Target gateset used for compiling circuits to √iSWAP + 1-q rotations + measurement gates."""
```
Why measurement gates ?

tanujkhattar replied:
Measurement gates will be "accepted" by the gatesets if they are part of the circuit. So, if you have a `PauliMeasurementGate` in your circuit and you call `cirq.optimize_for_target_gateset(c, gateset = cirq.SqrtIswapTargetGateset())`, it will decompose the pauli measurement into single qubit rotations + `cirq.Measurement` gates because pauli measurements are not part of the gateset but measurement gates are.

MichaelBroughton commented on cirq-core/cirq/transformers/target_gatesets/sqrt_iswap_gateset.py:
```diff
@@ -0,0 +1,281 @@
+            atol: A limit on the amount of absolute error introduced by the decomposition.
+            required_sqrt_iswap_count: When specified, the `decompose_to_target_gateset` will
+                decompose each operation into exactly this many sqrt-iSWAP gates even if fewer is
+                possible (maximum 3). A ValueError will be raised if this number is 2 or lower and
+                synthesis of the operation requires more.
```
Would an error also be raised if the number is 1 or lower and the operation synthesis requires more ?

tanujkhattar replied:
Yes.

MichaelBroughton reviewed (commented):
We are packing a lot into this PR and it's getting a little hard for me to follow test changes with deprecations woven in as well, would you mind splitting out the deprecation into a seperate PR ? Also why do we mention the support of measurement gates in this transformer it doesn't look like we are inserting measurements as a part of the circuit transformation process ?

tanujkhattar commented:
@MichaelBroughton As discussed offline, I've removed deprecations from this PR and they will follow up in the next PR once this is checked in. 

I've also abstracted out the parameterized gate compilation logic to a `cirq.parameterized_2q_op_to_sqrt_iswap_operations` analytical method and used it in `cirq.SqrtIswapTargetGateset`. Also added a few more tests that demonstrate the power of implicit decompose + parameterized compilation to support compilation of parameterized gates which are not directly supported by the analytical method. 

The PR is ready for a re-review! PTAL

MichaelBroughton commented on cirq-core/cirq/transformers/target_gatesets/sqrt_iswap_gateset.py:
```diff
@@ -0,0 +1,111 @@
+            f')'
+        )
+
+    def _value_equality_values_(self) -> Any:
```
Does this need a corresponding decorator on the top of the class to auto generate other equalities ?

tanujkhattar replied:
Equalities are auto generated because the class derives from `cirq.Gateset`, which is decorated with `@value.value_equality()`.

Added additional tests to assert this.

MichaelBroughton commented on cirq-core/cirq/transformers/analytical_decompositions/two_qubit_to_sqrt_iswap.py:
```diff
@@ -32,6 +33,201 @@
+            of `cirq.SQRT_ISWAP`.
+
+    Returns:
+        A parameterized `cirq.OP_TREE` implementing `op` using only `cirq.SQRT_ISWAP`
+        (or `cirq.SQRT_ISWAP_INV`) and parameterized single qubit rotations OR
+        None or NotImplemented if decomposition of `op` is not known.
```
Is it an op_tree or a DecomposeResult ? (from fxn return type hint)

tanujkhattar replied:
`DecomposeResult` is `Union[cirq.OP_TREE, None, NotImplementedType]` -- the docstring covers all 3 cases.

MichaelBroughton reviewed (approved):
LGTM after nits.
